runtime/near-vm-runner/src/wasmer_runner.rs

```diff
@@ -342,20 +331,44 @@ impl crate::runner::VM for Wasmer0VM {
             current_protocol_version,
         );
 
-        // TODO: charge this before module is loaded
-        if logic.add_contract_loading_fee(code.code().len() as u64).is_err() {
-            let err = VMError::FunctionCallError(FunctionCallError::HostError(
-                near_vm_errors::HostError::GasExceeded,
-            ));
-            return VMResult::abort(logic, err);
+        let result =
+            logic.before_code_loading(method_name, current_protocol_version, code.code().len());
+        if result.is_err() {
+            return VMResult::abort(logic, result.unwrap_err());
+        }
+
+        // TODO: consider using get_module() here, once we'll go via deployment path.
+        let module = cache::wasmer0_cache::compile_module_cached_wasmer0(code, &self.config, cache);
+        let module = match into_vm_result(module) {
+            Ok(x) => x,
+            // Note on backwards-compatibility: This error used to be an error
+            // without result, later refactored to NOP outcome. Now this returns
+            // an actual outcome, including gas costs that occurred before this
+            // point. This is compatible with earlier versions because those
+            // version do not have gas costs before reaching this code. (Also
+            // see `test_old_fn_loading_behavior_preserved` for a test that
+            // verifies future changes do not counteract this assumption.)
+            Err(err) => return VMResult::abort(logic, err),
+        };
+
+        let result = logic.after_code_loading(current_protocol_version, code.code().len());
+        if result.is_err() {
+            return VMResult::abort(logic, result.unwrap_err());
         }
 
         let import_object =
             imports::wasmer::build(memory_copy, &mut logic, current_protocol_version);
 
         if let Err(e) = check_method(&module, method_name) {
-            // TODO: This should return an outcome to account for loading cost
-            return VMResult::nop_outcome(e);
+            if checked_feature!(
+                "protocol_feature_fix_contract_loading_cost",
+                FixContractLoadingCost,
+                current_protocol_version
+            ) {
+                return VMResult::abort(logic, e);
+            } else {
+                return VMResult::nop_outcome(e);
+            }
```

Comment:
Would it be worthwhile perhaps to abstract this pattern into a method? It gets repetitive fast, and I don't think there's much more value in open coding it vs something like

```rust
VMResult::abort_but_nop_outcome_in_old_protocol(protocol_version, logic, e)
```